src/modm/driver/radio/sx1276_impl.hpp

```diff
@@ -0,0 +1,362 @@
+
+template <typename SpiMaster, typename Cs>
+modm::ResumableResult<void>
+Sx1276<SpiMaster, Cs>::transmit(uint8_t* data, uint8_t length)
+{
+    RF_BEGIN();
+
+    RF_CALL(changeMode(ModemMode::STDBY));
+
+    RF_CALL(writeRegister(Sx1276Register::PayloadLength,length));
+    RF_CALL(writeRegister(Sx1276Register::FifoTxBaseAddr,0x00));
+    RF_CALL(writeRegister(Sx1276Register::FifoAddrPtr,0x00));
+
+    RF_WAIT_UNTIL(this->acquireMaster());
+
+    buffer[0] = 0x80 | static_cast<uint8_t>(Sx1276Register::Fifo);
+
+	Cs::reset();
+
+    RF_CALL(SpiMaster::transfer(buffer,nullptr,1));
+    RF_CALL(SpiMaster::transfer(data,nullptr,length));
+
+	RF_WAIT_UNTIL(this->releaseMaster());
+	Cs::set();
+
+    RF_CALL(changeMode(ModemMode::TX));
+
+    RF_END();
+}
+
+// -----------------------------------------------------------------------------
+
+template <typename SpiMaster, typename Cs>
+modm::ResumableResult<void>
+Sx1276<SpiMaster, Cs>::enableListening()
+{
+    RF_BEGIN();
+    //enable the rx interrupts
+    irqMaskShadow.reset(Interrupts::RX_DONE);
+
+    RF_CALL(writeRegister(Sx1276Register::FifoAddrPtr,0x00));
+    RF_CALL(writeRegister(Sx1276Register::FifoRxBaseAddr,0x00));
+
+    RF_CALL(writeRegister(Sx1276Register::IrqFlagsMask,irqMaskShadow.value));
+    RF_CALL(writeRegister(Sx1276Register::IrqFlags, 0xFF));
+
+    RF_CALL(changeMode(ModemMode::RXCONTINOUS));
+    RF_END();
+}
+
+// -----------------------------------------------------------------------------
+
+template <typename SpiMaster, typename Cs>
+modm::ResumableResult<void>
+Sx1276<SpiMaster, Cs>::disableListening()
+{
+    RF_BEGIN();
+
+    irqMaskShadow.set(Interrupts::RX_DONE);
+    RF_CALL(writeRegister(Sx1276Register::IrqFlagsMask,irqMaskShadow.value));
+
+    RF_CALL(changeMode(ModemMode::SLEEP));
+
+    RF_END();
+}
+
+// -----------------------------------------------------------------------------
+
+template <typename SpiMaster, typename Cs>
+modm::ResumableResult<uint8_t>
+Sx1276<SpiMaster, Cs>::readPacket(uint8_t* data, uint8_t maxLength)
+{
+    RF_BEGIN();
+
+    irqFlags.value = RF_CALL(readRegister(Sx1276Register::IrqFlags));
+    RF_CALL(writeRegister(Sx1276Register::IrqFlags, 0xFF));
+
+    lastPacketSize = 0;
+
+    if(irqFlags & Interrupts::RX_DONE)
+    {
+        lastPacketSize = RF_CALL(readRegister(Sx1276Register::RxNbBytes));
+        RF_WAIT_UNTIL(this->acquireMaster());
+
+        buffer[0] = static_cast<uint8_t>(Sx1276Register::Fifo);
+
+        Cs::reset();
+
+        RF_CALL(SpiMaster::transfer(buffer,nullptr,1));
+        if(lastPacketSize > maxLength)
+        {
+            MODM_LOG_ERROR<<"Read buffer is too small, packet discarded!"<<modm::endl;
+            //read it out anyway to clean the fifo
+            RF_CALL(SpiMaster::transfer(nullptr,nullptr,lastPacketSize));
+        }
+        else
+        {
+            RF_CALL(SpiMaster::transfer(nullptr,data,lastPacketSize));
+        }
+
+        RF_WAIT_UNTIL(this->releaseMaster());
```

Comment:
This was meant to enable nesting of SPI routines, so that only the outer nesting resets the CS pin, not the inner ones. `this->releaseMaster()` returns true, if the nesting level is `0`.

Reply:
(I know this is not well documented).

Reply:
I first did the if condition around the releaseMaster but then sometimes the CS pin didn't get set to high. I guess I called it from a lower nesting level. I have to investigate this further